RecoBTag/ONNXRuntime/plugins/ParticleTransformerAK4SonicJetTagsProducer.cc

```diff
@@ -0,0 +1,306 @@
+  if (max_n_cpf == 0 && max_n_npf == 0 && max_n_vtx == 0) {
+    client_->setBatchSize(0);
+    skippedInference_ = true;
+    return;
+  }
+    
+  // all the jets in the same event will fill up the same amount of n_cpf, n_npf, n_vtx and send to server
+  max_n_cpf = std::clamp(max_n_cpf, (unsigned int)0, (unsigned int)25);
+  max_n_npf = std::clamp(max_n_npf, (unsigned int)0, (unsigned int)25);
+  max_n_vtx = std::clamp(max_n_vtx, (unsigned int)0, (unsigned int)5);
+
+  for (unsigned igroup = 0; igroup < input_names_.size(); ++igroup) {
+    const auto &group_name = input_names_[igroup];
+    auto &input = iInput.at(group_name);
+    unsigned target = 0;
+    
+    if (igroup == kChargedCandidates || igroup == kChargedCandidates4Vec) target = std::max((unsigned int)1, max_n_cpf);
+    else if (igroup == kNeutralCandidates || igroup == kNeutralCandidates4Vec) target = std::max((unsigned int)1, max_n_npf);
+    else if (igroup == kVertices || igroup == kVertices4Vec) target = std::max((unsigned int)1, max_n_vtx);
+    
+    input.setShape(0, target);
+    auto tdata = input.allocate<float>(true);
+    for (unsigned jet_n = 0; jet_n < tag_infos->size(); ++jet_n) {
+      const auto &taginfo = (*tag_infos)[jet_n];
+      const auto &features = taginfo.features();
+      auto &vdata = (*tdata)[jet_n];
+      // Loop through the n cpf, and in the case that n_cpf is smaller than max_n_cpf, add padding values to all features
+      if (igroup == kChargedCandidates) {
+        unsigned int n_cpf = features.c_pf_features.size();
+        n_cpf = std::clamp(n_cpf, (unsigned int)0,  (unsigned int)25);
+        for (unsigned int count = 0; count < n_cpf; count++) {
+          vdata.push_back(features.c_pf_features.at(count).btagPf_trackEtaRel);
+          vdata.push_back(features.c_pf_features.at(count).btagPf_trackPtRel);
+          vdata.push_back(features.c_pf_features.at(count).btagPf_trackPPar);
+          vdata.push_back(features.c_pf_features.at(count).btagPf_trackDeltaR);
+          vdata.push_back(features.c_pf_features.at(count).btagPf_trackPParRatio);
+          vdata.push_back(features.c_pf_features.at(count).btagPf_trackSip2dVal);
+          vdata.push_back(features.c_pf_features.at(count).btagPf_trackSip2dSig);
+          vdata.push_back(features.c_pf_features.at(count).btagPf_trackSip3dVal);
+          vdata.push_back(features.c_pf_features.at(count).btagPf_trackSip3dSig);
+          vdata.push_back(features.c_pf_features.at(count).btagPf_trackJetDistVal);
+          vdata.push_back(features.c_pf_features.at(count).ptrel);
+          vdata.push_back(features.c_pf_features.at(count).drminsv);
+          vdata.push_back(features.c_pf_features.at(count).vtx_ass);
+          vdata.push_back(features.c_pf_features.at(count).puppiw);
+          vdata.push_back(features.c_pf_features.at(count).chi2);
+          vdata.push_back(features.c_pf_features.at(count).quality);
```

Comment:
rather than duplicate these lists of features, extract the `make_inputs()` functions from https://github.com/cms-sw/cmssw/blob/master/RecoBTag/ONNXRuntime/plugins/ParticleTransformerAK4ONNXJetTagsProducer.cc into a helper class. the similar example for ParticleNet is https://github.com/cms-sw/cmssw/blob/master/RecoBTag/FeatureTools/interface/deep_helpers.h

Reply:
they are moved to interface/tensor_filler.h and src/tensor_filler.cc